unit_tests/test_rest_asset_client.py

```diff
@@ -0,0 +1,294 @@
+# pylint: disable=missing-function-docstring, no-member, too-many-public-methods
+# pylint: disable=too-many-arguments
+import uuid
+from unittest.mock import patch, MagicMock
+
+import pytest
+from gsy_e_sdk.constants import MAX_WORKER_THREADS
+
+from gsy_e_sdk.clients.rest_asset_client import RestAssetClient
+
+TEST_ASSET_UUID = str(uuid.uuid4())
+TEST_SIMULATION_ID = str(uuid.uuid4())
+TEST_AGGREGATOR_UUID = str(uuid.uuid4())
+TEST_TRANSACTION_ID = str(uuid.uuid4())
+
+TEST_DOMAIN_NAME = "test.domain@name.com"
+TEST_WEBSOCKETS_DOMAIN_NAME = "wss:test.domain@name.com"
+
+TEST_COMMAND_RESPONSE = {"command": "register",
+                         "transaction_id": TEST_TRANSACTION_ID,
+                         "registered": True
+                         }
+
+
+@pytest.fixture(name="mock_connections", autouse=True)
+def fixture_mock_connections(mocker):
+    # gsy_framework/client_connections/utils.py
+    mocker.patch("gsy_e_sdk.clients.rest_asset_client.retrieve_jwt_key_from_server")
+    mocker.patch("gsy_framework.client_connections.utils.RepeatingTimer")
+
+    mocker.patch("gsy_e_sdk.clients.rest_asset_client.WebsocketThread")
+    mocker.patch("gsy_e_sdk.clients.rest_asset_client.ThreadPoolExecutor")
+
+
+@pytest.fixture(name="mock_environment_use_functions")
+def fixture_mock_env_use_functions(mocker):
+    mocker.patch("gsy_e_sdk.clients.rest_asset_client.simulation_id_from_env",
+                 return_value=TEST_SIMULATION_ID)
+    mocker.patch("gsy_e_sdk.clients.rest_asset_client.domain_name_from_env",
+                 return_value=TEST_DOMAIN_NAME)
+    mocker.patch("gsy_e_sdk.clients.rest_asset_client.websocket_domain_name_from_env",
+                 return_value=TEST_WEBSOCKETS_DOMAIN_NAME)
+
+
+@pytest.fixture(name="mock_transaction_id")
+def mock_transaction_id_and_timeout_blocking(mocker):
+    mocker.patch("gsy_framework.client_connections.utils.uuid.uuid4",
+                 return_value=TEST_TRANSACTION_ID)
```

Comment:
Can you please explain why this is needed? Why mocking a standard library feature?

Reply:
In this case I mock it to make assertions using the id, forcing it to return a specific id previously generated. 

For example, as a side effect on register (due to `_post_request() `private method), a transaction id is generated and added to a dictionary used to post a request.

As the `post_request` message must be sent for the `rest_asset_client` to work properly, I made an assertion this being called with some explicit arguments. One of which is a dictionary containing a particular `transaction_id`. If I were to let `uuid4()` to generate a random id, there would be no way (AFAIK) for me to know this and assert if the post was made with the expected transaction_id or not.

However, if I assume `uuid4()` works properly (which it should) I can safely mock it to return a convenient uuid which I know beforehand, to assert that `post_request` was called with a dictionary containing this id.

Reply:
I see. Thanks a lot for the information!